### PR TITLE
Bug 1177044 - [Contacts][NGA] Move view() and utility() from contacts.js to a separate class

### PR DESCRIPTION
**Pull request by jpruden92**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=552603cded052b3b8eec86bc31cf33d3803eded0

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bbc4f1b5eb12a6e4ca3aa9fa4136e82a1515c098

File: apps/communications/contacts/js/loader.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    },
+
+    get SHARED_UTILS_PATH() {
+      return SHARED_UTILS_PATH;
+    }
+
+  };
+
+  exports.Loader = Loader;
+}(window));
```
Comment (borjasalguero) [inline apps/communications/contacts/js/loader.js]: Add one more line here

File: apps/communications/contacts/test/unit/cache_eviction_test.js [hunk 2]
```diff
@@ -75,6 +77,8 @@ suite('Contacts', function() {
     realNavigationStack = window.navigationStack;
     window.navigationStack = MockNavigationStack;
 
+    window.Loader = MockLoader;
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/cache_eviction_test.js]: You need to clear this in the `suiteTeardown`

File: apps/communications/contacts/test/unit/contacts_bulk_delete_test.js [hunk 3]
```diff
@@ -36,6 +37,7 @@ suite('contacts_bulk_delete.js', function() {
       window.utils = {};
     }
     realOverlay = window.utils.overlay;
+    window.Loader = MockLoader;
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/contacts_bulk_delete_test.js]: Clean this in `suiteTeardown`

File: apps/communications/contacts/test/unit/export/generic_export_test.js [hunk 4]
```diff
@@ -80,6 +82,8 @@ suite('Contacts Exporter', function() {
     real_ = window._;
     window._ = navigator.mozL10n.get;
 
+    window.Loader = MockLoader;
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/export/generic_export_test.js]: Same.

Comment (borjasalguero) [inline apps/communications/contacts/test/unit/export/generic_export_test.js]: Actually you can use `MocksHelper` in [1] and you have a more readable code.

[1] https://github.com/mozilla-b2g/gaia/pull/30700/files#diff-3407ab06d83e82e2d919ccbdd7ccb3eeR49

File: apps/communications/contacts/test/unit/mock_loader.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+/* exported MockLoader */
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/mock_loader.js]: Move this to the first line of the code (before `user strict`)

File: apps/communications/contacts/test/unit/views/settings_test.js [hunk 6]
```diff
@@ -116,6 +119,8 @@ suite('Contacts settings >', function() {
     window.utils = window.utils || {};
     window.utils.cookie = MockCookie;
 
+    window.Loader = MockLoader;
```
Comment (borjasalguero) [inline apps/communications/contacts/test/unit/views/settings_test.js]: It's the same as before. Fix all of them and we are done!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b3ec4fac95ec845b2564f5aea9d3c128ea6fe1f6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c430d97f8abd154b2c06260b44d56f1157e7aa57

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7ab3f4896149e741f0fc41f8947b3d47093ddeb7

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=09d15ca04f14872df5f0dff71c8eb25f2502ed45

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f30c73cc3d820c79300db573503c0b93533ac75f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b2caead80163f0f34e0564395e6624f6aa31f0a3

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f528be3dd6981d59ca2cd758ebeb0831a7d18183

File: apps/communications/contacts/js/contacts.js [hunk 7]
```diff
@@ -23,14 +24,8 @@ var COMMS_APP_ORIGIN = location.origin;
...
 var Contacts = (function() {
-  var SHARED = 'shared';
-  var SHARED_PATH = '/' + SHARED + '/' + 'js';
 
-  var SHARED_UTILS = 'sharedUtilities';
-  var SHARED_UTILS_PATH = SHARED_PATH + '/contacts/import/utilities';
-
-  var SHARED_CONTACTS = 'sharedContacts';
-  var SHARED_CONTACTS_PATH = SHARED_PATH + '/' + 'contacts';
+  var SHARED_UTILS_PATH = '/shared/js/contacts/import/utilities';
```
Comment (borjasalguero) [inline apps/communications/contacts/js/contacts.js]: Why dont we remove this? Let's use the complete path, or use something as `Loader.sharedPath` (due to this is replicated in `Loader`) for retrieving it (exposing it as a `get` would be fine).

File: apps/communications/contacts/js/contacts.js [hunk 8]
```diff
@@ -219,7 +214,9 @@ var Contacts = (function() {
 
     // We need to return the promise here for testing purposes
     return addAsyncScripts().then(() => {
-      checkUrl();
+      Loader.utility('Confirm', function viewLoaded() {
```
Comment (borjasalguero) [inline apps/communications/contacts/js/contacts.js]: Why do you need to load `Confirm` before checkUrl? Could we load this within `checkUrl` if needed just in the case we use it?

File: apps/communications/contacts/js/loader.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * Specifies dependencies for resources
+   * E.g., mapping Facebook as a dependency of views
+   */
+  var dependencies = {
+    views: {
+      Settings: Contacts.loadFacebook, //FIX: Should extract FB from Contacts.
+      Details: Contacts.loadFacebook,
+      Form: Contacts.loadFacebook,
```
Comment (borjasalguero) [inline apps/communications/contacts/js/loader.js]: Why do we need to keep `loadFacebook`?

File: apps/communications/contacts/js/loader.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
+(function(exports) {
```
Comment (arcturus) [inline apps/communications/contacts/js/loader.js]: nit: add jsdoc explaining what this class do

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a0236e84bea4f5fcf009237a51e5cecb796ab679

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2a6fb025b0287c8eeecdf4761839e648c3e231b9

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4ba2be894310ee945b8d8b7e441872cf643dd340

**Pull request merged**